### PR TITLE
hubble-relay: Add node status message

### LINKED ISSUE #11360: api/observer: include the number of peers that answered the request in GetFlows() response
In a multi-node context, such as when [`GetFlows` from the `Observer` service](https://github.com/cilium/cilium/blob/6899f9146ac1fb8195743cc9eb684d0f16b49594/api/v1/observer/observer.proto#L24) is implemented by hubble-relay,  the API lacks  a way to inform the caller when (at least) a peer failed to answer the request. With a large enough cluster, the probability of at least one peer to fail answering a  `GetFlows()` request is high thus one can't simply error out the entire request in such a scenario.

This is a proposal to add a field to the [`GetFlowsResponse`](https://github.com/cilium/cilium/blob/6899f9146ac1fb8195743cc9eb684d0f16b49594/api/v1/observer/observer.proto#L90) message type to inform about the number of peers that answered the request (could be in a form of `x/N` or 2 fields or else). Such a change should be backward compatible.

### PR DESCRIPTION
**Pull request by gandro**
This adds a status message to the observer service implementation
of hubble-relay. The status message is added as a new variant of
the GetFlowsResponse, therefore older clients will silently ignore
it.

The new node status event is defined as follow:

```protobuf
// NodeStatusEvent is a message sent by hubble-relay to inform clients about
// the state of a particular node.
message NodeStatusEvent {
    // state_change contains the new node state
    NodeState state_change = 1;
    // node_names is the list of nodes for which the above state changes applies
    repeated string node_names = 2;
    // message is an optional message attached to the state change (e.g. an
    // error message). The message applies to all nodes in node_names.
    string message = 3;
}

enum NodeState {
    // UNKNOWN_NODE_STATE indicates that the state of this node is unknown.
    UNKNOWN_NODE_STATE = 0;
    // NODE_CONNECTED indicates that we have established a connection
    // to this node. The client can expect to observe flows from this node.
    NODE_CONNECTED = 1;
    // NODE_UNAVAILABLE indicates that the connection to this
    // node is currently unavailable. The client can expect to not see any
    // flows from this node until either the connection is re-established or
    // the node is gone.
    NODE_UNAVAILABLE = 2;
    // NODE_GONE indicates that a node has been removed from the
    // cluster. No reconnection attempts will be made.
    NODE_GONE = 3;
    // NODE_ERROR indicates that a node has reported an error while processing
    // the request. No reconnection attempts will be made.
    NODE_ERROR = 4;
}
```

The implementation makes sure to send down the initial node status
for all nodes participating in the request first, i.e. before any flows.

The node state `GONE` is not used in this initial version, as the
current implementation of the peer management in hubble-relay does not
inform the running GetFlows request about the removal of a node.
We mark any failed node as `UNAVAILABLE` for now.

**Review per commit** as this PR also performs a bit of cleanup.

Fixes: #11360

### CODE REVIEW

File: api/v1/observer/observer.proto
Comment (rolinh) [inline api/v1/observer/observer.proto]: Thinking: this message definition seems generic enough that it could be reused in other gRPC APIs. I'm mostly thinking about the relay API proposal in #11026. With this in mind, what about moving the definition to `api/v1/relay/relay.proto`? And then maybe rename it to simply `Status` as `relay.RelayStatus` would clutter a bit.

File: pkg/hubble/relay/observer.go
Comment (rolinh) [inline pkg/hubble/relay/observer.go]: Does the node name makes sense in the context of hubble-relay? Well, I guess we have to fill this anyway and I see no better alternative.

File: pkg/hubble/relay/observer.go
Comment (rolinh) [inline pkg/hubble/relay/observer.go]: I guess that closing the flows channel here is assumed to be OK even though there are other routines that write into this channel because this goroutine stops only when the context is done, right? I'm not sure it's safe though because if this one returns on `ctx.Done()` before one of the other writers, I think there could be a chance for a tiny race that could lead to a panic (writing to closed channel).

**Review by rolinh — COMMENTED** [review]
Very nice improvement overall! I'm a little worried about the complexity of the `GetFlows` implementation. Eventually, it should be refactored so that it could be more readable/testable/debuggable (work for a follow-up PR though).

File: api/v1/observer/observer.proto
Comment (gandro) [reply api/v1/observer/observer.proto]: Good suggestion!

File: pkg/hubble/relay/observer.go
Comment (gandro) [reply pkg/hubble/relay/observer.go]: You are correct. I have not considered the case where the `ctx.Done()` is racy.

The basic "shutdown" procedure is intended to be as follows (in non-follow-mode):
 - Wait until each each peer in the errgroup finishes its request, this ensures we have collected all `failedPeers`. We indicate this by closing the `failedPeers` channel.
-  The status go-routine sees that it has received all failed peers, therefore it can generate the `RelayStatus` message with the complete list of peers. It then closes `flows`.

In follow-mode, we do not have to option to send back a final status message, since these kind of clients just disconnect. Therefore, we will eventually just get `io.EOF` from `stream.Send` which will cancel the context and tear down all go routines.

But yes, as it is currently we have the problem that we might write to closed channels. I will rework the logic to avoid this.

Comment (rolinh) [reply pkg/hubble/relay/observer.go]: Here is an idea to try and simplify this although I didn't try to implement the logic myself:
What about getting rid of the `failedPeers` channel altogether and leverage your new `RelayStatus` message instead? So when a peer fails, instead of sending `p.Name` in the `failedPeers` chan, you just send a `GetFlowsResponse` with a `RelayStatus` into the flows channel with information about the peer that failed. These status results can then be aggregated by the receiver and sent when appropriate (eg when the timer ticks); it's just a matter of checking for them when reading from the `flows` channel. This way you don't have to create an additional goroutine and can leave the tear-down logic as it was before.

File: api/v1/relay/relay.proto
Comment (rolinh) [inline api/v1/relay/relay.proto:33]: Since the enum itself is called `NodeState`, I don't think it's necessary to repeat `NODE_STATE` in each state name.
They could just be `UNKNOWN`, `CONNECTED`, `UNAVAILABLE` and `GONE`.

File: pkg/hubble/relay/observer.go
Comment (rolinh) [inline pkg/hubble/relay/observer.go:69]: As this is the only place where `sortedFlows` is written to, what about changing the prototype to:
```go
func sortFlows(
	ctx context.Context,
	flows <-chan *observerpb.GetFlowsResponse,
	qlen int,
	bufferDrainTimeout time.Duration,
) <-chan *observerpb.GetFlowsResponse {...}
```
And have `sortFlows` be responsible of creating and closing the channel (this requires spawning a goroutine in the func body rather than leaving this work to the caller)? I think the code would look a little cleaner from the caller's perspective as instead of this:
```go
       sortedFlows := make(chan *observerpb.GetFlowsResponse, qlen)
        go func() {
                defer close(sortedFlows)
                // sortFlows returns either when flows is closed or ctx expires
                sortFlows(ctx, flows, qlen, s.opts.BufferDrainTimeout, sortedFlows)
        }()
```
You'd have this:
```go
sortedFlows := sortFlows(ctx, flows, qlen, s.opts.BufferDrainTimeout)
```

**Review by rolinh — COMMENTED** [review]
I think the changes you made to the API proposal are good. However, I need to experiment/play a bit more with the API in order to understand what could be missing or improved before giving a final approval.

File: api/v1/relay/relay.proto
Comment (gandro) [reply api/v1/relay/relay.proto:33]: UNKNOWN needs to be prefixed, because protobuf does not have enum namespaces and will therefore collide with any other enum which will have the UNKNOWN variant. But I guess for the others, we can do that.

Comment (gandro) [reply api/v1/relay/relay.proto:33]: I have shortened the variants to `NODE_xxx`. I think a prefix is still useful, since `relay.proto` will gain quite a few other enums that maybe also want to make use of variants like `ERROR`.

**Comment by gandro** [comment]
I have added a `NODE_ERROR` status to the API. It is used to report errors which may occur during a request. 

Because certain errors occur on multiple nodes simultaneously (e.g. invalid request parameters), I have also added a function to coalesce duplicate errors.

**Comment by gandro** [comment]
test-me-please

File: pkg/hubble/relay/relayoption/defaults.go
Comment (rolinh) [inline pkg/hubble/relay/relayoption/defaults.go]: Was this value empirically chosen? It seems a little short to me.

**Review by rolinh — COMMENTED** [review]
I like the `NODE_ERROR` change, very welcome! As a follow-up to this PR, I wonder if we should not also record the last error for every peer along with the corresponding timestamp so that we could then serve this information via a corresponding note status grpc service.

File: pkg/hubble/relay/relayoption/defaults.go
Comment (gandro) [reply pkg/hubble/relay/relayoption/defaults.go]: For small cluster sizes, 3 seconds seems to work nicely: It's long enough such that each node will have a chance to report their error, but short enough so the clients don't have too long for errors to be reported.

But I actually don't know what a good window size would be for larger clusters. Too short and we end up producing unnecessary traffic, too long and we risk merging uncorrelated errors (and clients have to wait long for errors as well).

File: api/v1/relay/relay.proto
Comment (rolinh) [inline api/v1/relay/relay.proto]: I think it should be pluralized. This already confused me while reviewing https://github.com/cilium/hubble/pull/260 as I thought that `NodeName` was just a node name when in fact it's a list of node names.

File: pkg/hubble/relay/relayoption/defaults.go
Comment (gandro) [reply pkg/hubble/relay/relayoption/defaults.go]: Bumped to 10sec

**Comment by gandro** [comment]
test-me-please

File: pkg/hubble/relay/observer.go
Comment (rolinh) [inline pkg/hubble/relay/observer.go:125]: Why only if `s.Code() == codes.Unknown`?

File: pkg/hubble/relay/observer.go
Comment (rolinh) [inline pkg/hubble/relay/observer.go:196]: As I understand it, there are no node names de-dup logic, right?

File: pkg/hubble/relay/observer.go
Comment (gandro) [reply pkg/hubble/relay/observer.go:125]: It's mirroring the logic we have on the client-side: 

https://github.com/cilium/hubble/blob/9bca65beb242f467d869f044f54b58cf26e14c88/cmd/observe/observe.go#L90-L94

Unknown gRPC codes are errors thrown by the Hubble Observer for application-level errors such as `invalid IP address in filter`. Stripping out the `grpc error: code = Unknown desc =` makes this easier to understand for users, showing them that this is a problem with the request, not with grpc. 

All other codes (e.g. Unimplemented) tend to be protocol-level errors, therefore it makes sense to preserve the information that this is coming from the gRPC stack.

File: pkg/hubble/relay/observer.go
Comment (gandro) [reply pkg/hubble/relay/observer.go:196]: Yes. However, as it is implemented right now, a node can throw an error only once during a request. There cannot be any duplicate nodes to begin with.

**Comment by gandro** [comment]
retest-4.9

**Comment by gandro** [comment]
retest-4.19

File: pkg/hubble/relay/observer.go
Comment (rolinh) [reply pkg/hubble/relay/observer.go:196]: Ah, right because there is no retry logic yet.

**Review by rolinh — APPROVED** [review]
I've tested this quite extensively and I'm pretty happy with this new API change and implementation. Well done!

**Comment by gandro** [comment]
retest-runtime

**Comment by gandro** [comment]
retest-4.19

**Comment by gandro** [comment]
The runtime failure is a known flake: #10838.

**Review by aanm — APPROVED** [review]

**Pull request merged**
